File: 2021/d2/rust/src/main.rs
use std::path::Path;
use std::fs::File;
use std::io::Read;

fn locate1(lines: &Vec<&str>) -> (usize, usize) {
    let mut position = 0;
    let mut depth = 0;

    for line in lines {
        let mut split = line.split_ascii_whitespace();
        let direction = split.next().expect("no direction");
        let amount: usize = split.next().expect("no amount").parse().expect("amount couldn't be parsed");

        match direction {
            "forward" => position += amount,
            "down" => depth += amount,
            "up" => depth -= amount,
            _ => panic!("unknown direction: {}", direction)
        }
    }

    (position, depth)
}

fn locate2(lines: &Vec<&str>) -> (usize, usize) {
    let mut position = 0;
    let mut depth = 0;
    let mut aim = 0;

    for line in lines {
        let mut split = line.split_ascii_whitespace();
        let direction = split.next().expect("no direction");
        let amount: usize = split.next().expect("no amount").parse().expect("amount couldn't be parsed");

        match direction {
            "forward" => {
                position += amount;
                depth += amount * aim;
            },
            "down" => aim += amount,
            "up" => aim -= amount,
            _ => panic!("unknown direction: {}", direction)
        }
    }

    (position, depth)
}

fn main() {
    let mut args = std::env::args();

    let input_filename = args.nth(1).unwrap_or_else(|| {
        eprintln!("No input file specified");
        eprintln!("Usage: ./run <input filename>");
        std::process::exit(1);
    });

    let mut input_file = File::open(Path::new(&input_filename)).expect("failed to open input file");
    let mut input = String::new();
    input_file.read_to_string(&mut input).expect("failed to read from file");

    let lines = input.lines()
        .collect();

    let (position, depth) = locate1(&lines);

    println!("part 1: position: {}, depth: {} -- multiplied: {}", position, depth, position * depth);

    let (position, depth) = locate2(&lines);

    println!("part 2: position: {}, depth: {} -- multiplied: {}", position, depth, position * depth);
}

File: 2021/d2/python/main.py
import sys

def parse_command(line):
    l = line.split(' ')
    l[1] = int(l[1])
    return l


def drive_direct(commands):
    x = 0
    y = 0

    for command, amount in commands:
        if command == "forward":
            x += amount
        elif command == "up":
            y -= amount
        elif command == "down":
            y += amount
        else:
            raise Exception(f"unrecognized command: {command}")

    return x * y


def drive_aim(commands):
    x = 0
    y = 0
    aim = 0

    for command, amount in commands:
        if command == "forward":
            x += amount
            y += aim * amount
        elif command == "up":
            aim -= amount
        elif command == "down":
            aim += amount
        else:
            raise Exception(f"unrecognized command: {command}")

    return x * y


if __name__ == '__main__':
    if len(sys.argv) < 2:
        raise Exception("filename not speciifed, specify filename as first argument")

    filename = sys.argv[1]
    lines = None

    with open(filename) as f:
        lines = [parse_command(line.strip()) for line in f]

    if not lines:
        raise Exception("no lines found")

    print("part 1: " + str(drive_direct(lines)))
    print("part 2: " + str(drive_aim(lines)))
